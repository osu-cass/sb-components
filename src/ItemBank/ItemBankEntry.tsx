import * as React from "react";
import {
  Accordion,
  ItemRevisionModel,
  NamespaceModel,
  SectionModel,
  ItemEntryTable,
  CsvEntry
} from "@src/index";

export interface ItemBankEntryProps {
  deleteItem: (item: number) => void;
  clearItems: () => void;
  submitItems: (items: ItemRevisionModel[]) => void;
  namespaces: NamespaceModel[];
  sections: SectionModel[];
  items: ItemRevisionModel[];
}

export interface ItemBankEntryState {
  csvIsOpen: boolean;
  itemsEntryIsOpen: boolean;
}

export class ItemBankEntry extends React.Component<
  ItemBankEntryProps,
  ItemBankEntryState
> {
  constructor(props: ItemBankEntryProps) {
    super(props);
    this.state = {
      csvIsOpen: false,
      itemsEntryIsOpen: false
    };
  }

  onAccordionToggle = (accordionType: "csv" | "table") => {
    const { csvIsOpen, itemsEntryIsOpen } = this.state;
    if (accordionType === "csv") {
      this.setState({ csvIsOpen: !csvIsOpen });
    } else {
      this.setState({ itemsEntryIsOpen: !itemsEntryIsOpen });
    }
  };

  onCsvApply = () => {
    this.setState({
      csvIsOpen: false,
      itemsEntryIsOpen: true
    });
  };

  renderCsvEntry() {
    const { namespaces, submitItems } = this.props;

    return (
      <Accordion
        accordionTitle="CSV Entry"
        isOpen={this.state.csvIsOpen}
        toggleExpand={() => this.onAccordionToggle("csv")}
      >
        <div className="csv-entry-container section">
          <div className="csv-entry-wrapper">
            <CsvEntry
              namespaces={namespaces}
              onApply={this.onCsvApply}
              onItemsUpdate={submitItems}
              itemRows={this.props.items}
            />
          </div>
        </div>
      </Accordion>
    );
  }

  renderTableEntry() {
    const {
      items,
      clearItems,
      namespaces,
      sections,
      deleteItem,
      submitItems
    } = this.props;

    return (
      <Accordion
        accordionTitle="Items Entry"
        isOpen={this.state.itemsEntryIsOpen}
        toggleExpand={() => this.onAccordionToggle("table")}
      >
        <ItemEntryTable
          itemRows={items}
          namespaces={namespaces}
          sections={sections}
          onDeleteItem={deleteItem}
          onClearItems={clearItems}
          onSubmit={submitItems}
        />
      </Accordion>
    );
  }

  render() {
    return (
      <div>
        {this.renderCsvEntry()}
        {this.renderTableEntry()}
      </div>
    );
  }
}
