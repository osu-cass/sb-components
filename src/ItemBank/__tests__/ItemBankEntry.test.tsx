import * as React from "react";
import { shallow } from "enzyme";
import { ItemBankEntry } from "../ItemBankEntry";
import {
  NamespaceModel,
  SectionModel,
  ItemRevisionModel
} from "../ItemBankModels";

const namespace: NamespaceModel = {
  hasBankKey: true,
  bankKey: 0,
  id: "12",
  name: "itemreviewviewer"
};

const section: SectionModel = {
  key: "187-3000",
  value: "math"
};

const item: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "asdge",
  isaap: ""
};

const items: ItemRevisionModel[] = [item];

const sections: SectionModel[] = [section];

const namespaces: NamespaceModel[] = [namespace];

describe("ItemBankEntry", () => {
  const wrapper = shallow(
    <ItemBankEntry
      updateItems={items => jest.fn()}
      namespaces={namespaces}
      sections={sections}
      items={items}
    />
  );

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
