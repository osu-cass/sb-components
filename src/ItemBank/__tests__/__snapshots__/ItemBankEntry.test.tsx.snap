// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemBankEntry matches snapshot 1`] = `
<div>
  <Accordion
    accordionTitle="CSV Entry"
    isOpen={false}
    toggleExpand={[Function]}
  >
    <div
      className="csv-entry-container section"
    >
      <div
        className="csv-entry-wrapper"
      >
        <CsvEntry
          onBlur={[Function]}
          onItemsUpdate={[Function]}
        />
      </div>
    </div>
  </Accordion>
  <Accordion
    accordionTitle="Items Entry"
    isOpen={false}
    toggleExpand={[Function]}
  >
    <ItemEntryTable
      itemRows={
        Array [
          Object {
            "bankKey": 3000,
            "isaap": "",
            "itemKey": 187,
            "revision": "asdge",
            "section": "math",
          },
        ]
      }
      namespaces={
        Array [
          Object {
            "bankKey": 0,
            "hasBankKey": true,
            "id": "12",
            "name": "itemreviewviewer",
          },
        ]
      }
      onItemsUpdate={[Function]}
      sections={
        Array [
          Object {
            "key": "187-3000",
            "value": "math",
          },
        ]
      }
    />
  </Accordion>
</div>
`;
